source/probability_theory_4_finite.Rmd

```diff
@@ -524,284 +481,178 @@ sampling mechanism as in Example 8-1, but now we record "yes" for a
-
-`           SUBTRACT a c d         `
-
-This operation pairs each shuffled female with a male and subtracts. If
-it is an original pairing (1/1, 2/2, etc.), the result will be a 0. The
-number of 0's indicates how many of the 10 males got paired up again
-with their original partner.
-
-`           COUNT d =0 e         `
-
-Count the number of 0's and put the result in e.
-
-`           SCORE e z         `
-
-Keep track of each trial result.
-
-End the trial, go back and repeat until all 1000 are complete.
-
-`           HISTOGRAM z         `
-
-Produce a histogram of the results.
+```{python, opts.label="py_ed"}
+N = 10000
+trial_results = np.zeros(N, dtype=int)
 
-![](images/12-Chap-8_007.png)
+# Assign numbers to each student, according to their pair, after the first
+# icebreaker
+birmingham = np.arange(10)
+berkeley = np.arange(10)
 
-From this histogram, we see that in about 3 percent of the trials did 4
-or more couples end up being re-paired with their own partners. We can
-calculate this proportion directly with RESAMPLING STATS:
+for i in range(N):
+    # Randomly shuffle the students from berkeley
+    # (We could also have shuffled both arrays---it does not alter the outcome)
+    shuffled_berkeley = np.random.permutation(berkeley)
 
-`           COUNT z >= 4 k         `
+    # Count in how many cases people landed with the same person as in the
+    # first round, and store in trial_results.
+    matches = np.count_nonzero(shuffled_berkeley == birmingham)
+    trial_results[i] = matches
 
-Determine how many trials had 4 or more males being matched with their
-partner after the shuffling. (Note that this is the same as the number
-of females being matched with their original partner.)
+# Count the number of times we got 4 or more people assigned to the same person
+k = np.count_nonzero(trial_results >= 4)
 
-`           DIVIDE k 1000 kk         `
+# Convert to a proportion.
+kk = k / N
 
-Convert to a proportion.
+# Print the result.
+print(kk)
+```
 
-`           PRINT kk         `
+**NOTE:** This seems to be 1.5 percent, instead of 3 now?
 
-Print the result.
+We see that in about 3 percent of the trials did 4 or more couples end
+up being re-paired with their own partners.  This can also be seen
+from the histogram:
 
-Note: The file "couples" on the Resampling Stats software disk contains
-this set of commands.
+```{python, opts.label="py_ed"}
+# Produce a histogram of trial results.
+plt.hist(trial_results, bins=range(10), align='left', density=True, rwidth=0.75)
+plt.title('Same pairs in round two');
+```
 
-## Example 8-6: Matching Hats: Another famous problem of this sort: The hat-checker at a restaurant mixes up the hats of a party of 6 men. What is the probability that at least one will get his own hat?
+## Example 8-6: Matching Hats: Another famous problem of this sort: The hat-checker at a restaurant mixes up the hats of a party of 6. What is the probability that at least one will get his own hat?
```

Comment:
This one a bit dated no?  Maybe coats instead?

Reply:
Or hats from Halloween costume?   Or hats from party of Santas clocking off after Christmas eve?

Reply:
What is this hat-swapping thing?!  Never heard of it.